### PR TITLE
Using protos for Asset Transfer in fabric.

### PR DESCRIPTION
**Pull request by sanvenDev**
1. Using protos for Asset Transfer in fabric.
2. Clean up `simpleassettransfer` and `simpleasset` chaincode.
3. Added new local query function to `simpleassettransfer` named `GetAssetPledgeDetails` to fetch asset pledge details locally (not for remote query).
4. Updated fabric-cli, added native fabric-cli commands for `claim` and `reclaim` instead of scripts.
5. Moved common code for interop call to helpers, and some fabric-cli bug fixes.
6. Updated docs as per new changes.

Edit:
1. Added workflow to automate deployment of docs as soon as `docs` folder is updated on PR merge.

### CODE REVIEW

File: core/network/fabric-interop-cc/libs/utils/utils.go [hunk 1]
```diff
@@ -151,40 +134,88 @@ func generateSHA256HashInHexForm(preimage string) string {
...
-	if id != "" && numUnits > 0 {
-        return "", fmt.Errorf("ambiguous pledge instruction: both asset ID and unit count provided")
+	if len(assetPledgeSerialized) == 0 {
+		return pledge, fmt.Errorf("empty asset pledge")
 	}
-	assetId := id           // Non-fungible asset
-	if assetId == "" {      // Fungible asset
-		assetId = generateAssetPledgeContractId(ctx, assetType, numUnits, owner, remoteNetworkId, recipientCert, expiryTimeSecs)
+	err = proto.Unmarshal([]byte(assetPledgeSerialized), pledge)
+    if err != nil {
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/libs/utils/utils.go:165]: Indentation seems to be off in this of-block.

File: core/network/fabric-interop-cc/libs/utils/utils.go [hunk 2]
```diff
@@ -151,40 +134,88 @@ func generateSHA256HashInHexForm(preimage string) string {
...
 		if err != nil {
 			return "", err
 		}
-		if (pledge.RemoteNetworkID == remoteNetworkId && pledge.RecipientCert == recipientCert && pledge.ExpiryTimeSecs == expiryTimeSecs) {
-			return assetId, nil
+		if (pledge.RemoteNetworkID == remoteNetworkId && pledge.Recipient == recipientCert && pledge.ExpiryTimeSecs == expiryTimeSecs) {
+			return pledgeId, nil
 		} else {
-			return "", fmt.Errorf("the asset %s has already been pledged", id)
+			return "", fmt.Errorf("the asset %s %s has already been pledged", assetType, assetIdOrQuantity)
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/libs/utils/utils.go:218]: Can you add `...%s with id %s   `

File: core/network/fabric-interop-cc/libs/utils/utils.go [hunk 3]
```diff
@@ -337,86 +387,117 @@ func ReclaimAsset(ctx contractapi.TransactionContextInterface, assetType, id, re
...
+	lookupPledgeBytes64, err := marshalAssetPledge(lookupPledge)
+	if err != nil {
+		return nil, pledgeBytes64, pledgeBytes64, err
+	}
+	
+	return lookupPledge.AssetDetails, lookupPledgeBytes64, pledgeBytes64, nil
+}
+
+// GetAssetPledgeDetails returns the asset pledge details for local network.
+func GetAssetPledgeDetails(ctx contractapi.TransactionContextInterface, pledgeId string) ([]byte, string, error) {
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/libs/utils/utils.go:433]: Isn't this function redundant, given we have `GetAssetPledgeStatus` above?

File: docs/docs/external/getting-started/test-network/ledger-initialization.md [hunk 4]
```diff
@@ -198,20 +198,29 @@ Prepare `fabric-cli` for configuration suitably as follows.
     - Set the `chaincode` attribute in each network to `simpleassettransfer`.
     - Set the `aclPolicyPrincipalType` attribute in `network2` to `ca`.
     - Otherwise, leave the default values unchanged.
+- Create `remote-network-config.json` file by copying `remote-network-config.json.template`. Use default values if Relays and Drivers are deployed in the host machines; else if they are deployed in docker, update as follows:
```
Comment (VRamakrishna) [inline docs/docs/external/getting-started/test-network/ledger-initialization.md:201]: Un-capitalize Relay, Driver. Capitalize docker. "machines" --> "machine".

File: samples/fabric/fabric-cli/.env.template [hunk 5]
```diff
@@ -5,4 +5,5 @@ LOCAL=true
 DEFAULT_APPLICATION_CHAINCODE=simplestate
 DEFAULT_APPLICATION_FUNC=Create
 CONFIG_PATH=PATH-TO-CONFIG-FILE
+REMOTE_CONFIG_PATH=PATH-TO-REMOTE-CONFIG-FILE
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/.env.template:8]: Move this line below the next to make it consistent with the instructions in the docs.

File: samples/fabric/fabric-cli/remote-network-config.json.template [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  "network2": {
+    "type": "fabric",
+    "relayEndpoint": "localhost:9083",
+    "channelName": "mychannel",
+    "chaincode": "simpleassettransfer"
+  },
+  "Corda_Network": {
+    "type": "corda",
+    "relayEndpoint": "localhost:9081",
+    "PartyAEndPoint": "localhost:10006",
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/remote-network-config.json.template:17]: Why is this keyword different from that in Line 23?

File: samples/fabric/fabric-cli/scripts/claimAsset.sh [hunk 7]
```diff
@@ -17,14 +17,17 @@ cd scripts
 USER_CERT_BASE64=$(node getAssetTransferViewAddress.js getusercert $2 alice)
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/scripts/claimAsset.sh:17]: Do we need this script anymore?

File: samples/fabric/fabric-cli/scripts/getAssetTransferViewAddress.js [hunk 8]
```diff
@@ -9,7 +9,7 @@ function getECertBase64(networkId, userId) {
     return Buffer.from(userIdJSON.credentials.certificate).toString('base64');
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/scripts/getAssetTransferViewAddress.js:9]: Do we need this script anymore?

File: samples/fabric/fabric-cli/scripts/reclaimAsset.sh [hunk 9]
```diff
@@ -16,11 +16,10 @@ cd scripts
 # Create a chaincode.json from the template file after replacing user IDs with certificates
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/scripts/reclaimAsset.sh:16]: Do we need this script anymore?

File: samples/fabric/fabric-cli/src/commands/asset/transfer/claim.ts [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    interopHelper
+} from '../../../helpers/helpers'
+import {
+    fabricHelper,
+    getUserCertBase64
+} from '../../../helpers/fabric-functions'
+
+import logger from '../../../helpers/logger'
+import * as dotenv from 'dotenv'
+dotenv.config({ path: path.resolve(__dirname, '../../../.env') })
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/src/commands/asset/transfer/claim.ts:26]: It'd be good to insert a blank link above this.

File: samples/fabric/fabric-cli/src/helpers/helpers.ts [hunk 11]
```diff
@@ -572,29 +632,157 @@ const generateViewAddress = async (
...
+            remoteNetConfig.chaincode + ':' + funcName + ':' + funcArgs.join(':');
+    } else if (remoteNetConfig.type == "corda") {
+        address = address + '/' + remoteNetConfig.PartyAEndPoint + '#' +
+            remoteNetConfig.flowPackage + "." + funcName + ":" + funcArgs.join(':');
+    } else {
+        logger.error(`Error: remote network ${remoteNetConfig.type} not supported.`)
+        throw new Error(`Error: remote network ${remoteNetConfig.type} not supported.`)
+    }
+    
+    return address
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/src/helpers/helpers.ts:694]: Some lines end in semicolons and others don't. Can we keep this consistent at least within files?

File: samples/fabric/simpleassettransfer/bondasset.go [hunk 12]
```diff
@@ -123,49 +140,76 @@ func (s *SmartContract) CreateAsset(ctx contractapi.TransactionContextInterface,
...
 	}
+	return asset, nil
+}
 
-	// Ensure that the client is the owner of the asset using app-specific logic
-	if !checkAccessToAsset(s, ctx, assetJSON) {
-		return fmt.Errorf("Canot pledge asset %s", id)
+// DeleteAsset deletes an given asset from the world state.
+func (s *SmartContract) DeleteAsset(ctx contractapi.TransactionContextInterface, assetType, id string) error {
+
```
Comment (VRamakrishna) [inline samples/fabric/simpleassettransfer/bondasset.go:158]: Remove whitespaces.

File: samples/fabric/simpleassettransfer/bondasset.go [hunk 13]
```diff
@@ -123,49 +140,76 @@ func (s *SmartContract) CreateAsset(ctx contractapi.TransactionContextInterface,
...
+// PledgeAsset locks an asset for transfer to a different ledger/network.
+func (s *SmartContract) PledgeAsset(ctx contractapi.TransactionContextInterface, assetType, id, remoteNetworkId, recipientCert string, expiryTimeSecs uint64) (string, error) {
+	// Check if asset is pledged already
+	bondAssetPledgeMap, err := getAssetPledgeIdMap(ctx, assetType, id)
+	if (err == nil) && (bondAssetPledgeMap.RemoteNetworkID == remoteNetworkId) && (bondAssetPledgeMap.Recipient == recipientCert) {
+		return bondAssetPledgeMap.PledgeID, nil
+	} else if (err == nil) {
+		return bondAssetPledgeMap.PledgeID, fmt.Errorf("asset %s is already pledged with pledgeId %s for different recipient", id, bondAssetPledgeMap.PledgeID)
+	}
+
```
Comment (VRamakrishna) [inline samples/fabric/simpleassettransfer/bondasset.go:176]: Remove whitespaces.

File: samples/fabric/simpleassettransfer/bondasset.go [hunk 14]
```diff
@@ -123,49 +140,76 @@ func (s *SmartContract) CreateAsset(ctx contractapi.TransactionContextInterface,
...
+		return bondAssetPledgeMap.PledgeID, fmt.Errorf("asset %s is already pledged with pledgeId %s for different recipient", id, bondAssetPledgeMap.PledgeID)
+	}
+	
+	// Read the asset (which internally checks access)
+	asset, err := s.ReadAsset(ctx, assetType, id)
 	if err != nil {
-		return err
+		return "", err
+	}
+
```
Comment (VRamakrishna) [inline samples/fabric/simpleassettransfer/bondasset.go:182]: Remove whitespaces.

File: samples/fabric/simpleassettransfer/bondasset.go [hunk 15]
```diff
@@ -311,64 +361,36 @@ func (s *SmartContract) GetAssetClaimStatus(ctx contractapi.TransactionContextIn
...
-	err = json.Unmarshal(assetJSON, &asset)
-	if err != nil {
-		return blankClaimJSON, err
+		return blankClaimBytes64, fmt.Errorf("failed to read asset record from world state: %v", err)
 	}
 	if asset.Owner != recipientCert {
-		return blankClaimJSON, nil      // Return blank
-	}
+		return blankClaimBytes64, nil      // Return blank
+	}
```
Comment (VRamakrishna) [inline samples/fabric/simpleassettransfer/bondasset.go:381]: Remove whitespaces at the end.

File: core/network/fabric-interop-cc/libs/utils/utils.go (hunk 3 above)
Comment (sanvenDev) [reply core/network/fabric-interop-cc/libs/utils/utils.go:433]: `GetAssetPledgeStatus`, I saw this function has lots of safety checks required during Asset Transfer protocol, which requires RecipientCert, RemoteNetworkID etc.

`GetAssetPledgeDetails`, is a simple query function for local network, to fetch all the details using just the `pledgeId`.

File: samples/fabric/fabric-cli/.env.template (hunk 5 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/.env.template:8]: Changed in docs, I remember maintaining this order somewhere with in fabric cli code as well.

File: samples/fabric/fabric-cli/remote-network-config.json.template (hunk 6 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/remote-network-config.json.template:17]: Its a typo, corrected.

File: samples/fabric/fabric-cli/scripts/getAssetTransferViewAddress.js (hunk 8 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/scripts/getAssetTransferViewAddress.js:9]: No, we don't, thought to keep it till my changes were finalized. Should I delete them?

File: samples/fabric/fabric-cli/src/commands/asset/transfer/claim.ts (hunk 10 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/src/commands/asset/transfer/claim.ts:26]: It's like this in every fabric-cli file which requires `.env`...

File: samples/fabric/simpleassettransfer/bondasset.go (hunk 13 above)
Comment (sanvenDev) [reply samples/fabric/simpleassettransfer/bondasset.go:176]: Keeping it here to not make it look cluttered...

File: samples/fabric/fabric-cli/scripts/reclaimAsset.sh (hunk 9 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/scripts/reclaimAsset.sh:16]: No, actually Initially I was trying to change with in this, so may be that's why these have come up in this PR.
I can delete the scripts if you are okay with it...

File: samples/fabric/fabric-cli/scripts/claimAsset.sh (hunk 7 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/scripts/claimAsset.sh:17]: No, we don't, actually Initially I was trying to change with in this, so may be that's why these have come up in this PR.
I thought to keep it till my changes were finalized. Should I delete them?

File: core/network/fabric-interop-cc/libs/utils/utils.go (hunk 3 above)
Comment (VRamakrishna) [reply core/network/fabric-interop-cc/libs/utils/utils.go:433]: Got it.

File: samples/fabric/fabric-cli/scripts/claimAsset.sh (hunk 7 above)
Comment (VRamakrishna) [reply samples/fabric/fabric-cli/scripts/claimAsset.sh:17]: Yes, you can delete them. If we ever need to refer to them, we can look at the revision history.

File: samples/fabric/fabric-cli/scripts/getAssetTransferViewAddress.js (hunk 8 above)
Comment (VRamakrishna) [reply samples/fabric/fabric-cli/scripts/getAssetTransferViewAddress.js:9]: Yes, you can delete them. If we ever need to refer to them, we can look at the revision history.

File: samples/fabric/fabric-cli/scripts/reclaimAsset.sh (hunk 9 above)
Comment (VRamakrishna) [reply samples/fabric/fabric-cli/scripts/reclaimAsset.sh:16]: Yes, you can delete them. If we ever need to refer to them, we can look at the revision history.

File: samples/fabric/fabric-cli/src/commands/asset/transfer/claim.ts (hunk 10 above)
Comment (VRamakrishna) [reply samples/fabric/fabric-cli/src/commands/asset/transfer/claim.ts:26]: OK, leave it as is.

**Review by VRamakrishna — APPROVED** [review]

**Pull request merged**